crates/boojum/src/gadgets/u256/mod.rs

```diff
@@ -351,6 +358,64 @@ impl<F: SmallField> UInt256<F> {
         });
         Self::from_le_bytes(cs, bytes)
     }
+
+    /// Finds the result of multiplying `self` by `other` mod `modulo`.
+    pub fn modmul<CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        other: &UInt256<F>,
+        modulo: &UInt256<F>,
+    ) -> UInt256<F> {
+        let a = self.witness_hook(cs)().unwrap_or_else(|| {
+            assert!(!<CS::Config as CSConfig>::WitnessConfig::EVALUATE_WITNESS);
+            U256::zero()
+        });
+
+        let b = other.witness_hook(cs)().unwrap_or_else(|| {
+            assert!(!<CS::Config as CSConfig>::WitnessConfig::EVALUATE_WITNESS);
+            U256::zero()
+        });
+
+        let m = modulo.witness_hook(cs)().unwrap_or_else(|| {
+            assert!(!<CS::Config as CSConfig>::WitnessConfig::EVALUATE_WITNESS);
+            U256::one()
+        });
+
+        let product = a.full_mul(b);
+
+        let (q, r) = match m.is_zero() {
+            true => (U512::zero(), U512::zero()),
+            false => product.div_mod(m.into()),
+        };
+
+        let q: U256 = q.try_into().unwrap();
+        let r: U256 = r.try_into().unwrap();
+
+        let q = UInt256::allocate(cs, q);
+        let r = UInt256::allocate(cs, r);
+
+        let mod_is_zero = Boolean::allocate(cs, m.is_zero());
+        let bool_true = Boolean::allocated_constant(cs, true);
+        let bool_false = Boolean::allocated_constant(cs, false);
+
+        let (_, m_ge_than_r) = r.overflowing_sub(cs, &modulo);
+        let m_ge_than_r = Boolean::conditionally_select(cs, mod_is_zero, &bool_true, &m_ge_than_r);
+        Boolean::enforce_equal(cs, &m_ge_than_r, &bool_true);
+
+        let lhs = self.widening_mul(cs, other, 8, 8);
+        let zero = UInt512::zero(cs);
+        let lhs = UInt512::conditionally_select(cs, mod_is_zero, &zero, &lhs);
+
+        let rhs = q.widening_mul(cs, &modulo, 8, 8);
+        let r_u512 = r.to_u512(cs);
+        let (rhs, overflow) = rhs.overflowing_add(cs, &r_u512);
+        Boolean::enforce_equal(cs, &overflow, &bool_false);
+
+        let are_equal = UInt512::equals(cs, &lhs, &rhs);
```

Comment:
Dont do this twice; write enforce_equal for UInt512

Reply:
No `enforce_equal` is implemented for any uint gadgets unfortunately